runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java

```diff
@@ -0,0 +1,201 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.spark.io;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.beam.runners.spark.translation.SparkRuntimeContext;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.io.Source;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.spark.Dependency;
+import org.apache.spark.InterruptibleIterator;
+import org.apache.spark.Partition;
+import org.apache.spark.SparkContext;
+import org.apache.spark.TaskContext;
+import org.apache.spark.api.java.JavaSparkContext$;
+import org.apache.spark.rdd.RDD;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Classes implementing Beam {@link Source} {@link RDD}s.
+ */
+public class SourceRDD {
+
+  /**
+   * A SourceRDD.Bounded reads input from a {@link BoundedSource} and creates a Spark {@link RDD}.
+   * This is the default way for the SparkRunner to read data from Beam's BoundedSources.
+   */
+  public static class Bounded<T> extends RDD<WindowedValue<T>> {
+    private static final Logger LOG = LoggerFactory.getLogger(SourceRDD.Bounded.class);
+
+    private final BoundedSource<T> source;
+    private final SparkRuntimeContext runtimeContext;
+    private final int numPartitions;
+
+    // to satisfy Scala API.
+    private static final scala.collection.immutable.List<Dependency<?>> NIL =
+        scala.collection.immutable.List.empty();
+
+    public Bounded(SparkContext sc,
+                   BoundedSource<T> source,
+                   SparkRuntimeContext runtimeContext) {
+      super(sc, NIL, JavaSparkContext$.MODULE$.<WindowedValue<T>>fakeClassTag());
+      this.source = source;
+      this.runtimeContext = runtimeContext;
+      // the input parallelism is determined by Spark's scheduler backend.
+      // when running on YARN/SparkDeploy it's the result of max(totalCores, 2).
+      // when running on Mesos it's 8.
+      // when running local it's the total number of cores (local = 1, local[N] = N,
+      // local[*] = estimation of the machine's cores).
+      // ** the configuration "spark.default.parallelism" takes precedence over all of the above **
+      this.numPartitions = sc.defaultParallelism();
+      checkArgument(this.numPartitions > 0, "Number of partitions must be greater than zero.");
```

Comment:
Well, it comes from the `SparkContext`, how it determines it - that's Spark's business.
Spark requires set all kinds of resource-related settings, it's not reasonable to validate them all.
If you run local/YARN/Mesos then you're probably aware of that, and if you explicitly mentioned parallelism conf. you're aware as well.